src/library.js

```diff
@@ -3532,6 +3532,46 @@ LibraryManager.library = {
     }
   },
 
+#if SHRINK_LEVEL == 0
+  // A mirror copy of contents of wasmTable in JS side, to avoid relatively
+  // slow wasmTable.get() call. Only used when not compiling with -Os or -Oz.
+  _wasmTableMirror: [],
```

Comment:
We've been using `$` prefixes to keep the symbols out of the native-linking symbol space which we certainly want to do for this symbol.

We don't have a good pattern for marking things as internal which I guess is the idea here.   Perhaps `$` + `_` if you prefer but I'd settle for just `$` until we come up with standard way to mark things as internal.

Reply:
> We've been using `$` prefixes to keep the symbols out of the native-linking symbol space which we certainly want to do for this symbol.

I think we discussed this at some point before. Let me try to revisit.

Historically, `$` was used predominantly for public JS APIs (`$GL`, `$FS`, `$Browser`, `$GLctx`, `$JSEvents`, etc.), as whatever is prepended with `$` pollutes the user JS namespace, which could cause conflict for developers. The presence of the underscore prefix has hinted towards an API being internal, and not to be relied on by JS developers, and the presence of `$` marked that the code was available for external JS to access, since it is exposed without prefixes.

I think when migrating to Wasm backend without C name mangling the intention of `$`  and `_` changed for some reason.

> We don't have a good pattern for marking things as internal

Prepending JS functions with a underscore `_` has always been the way to mark JS functions internal as far as I remember. JS linker prepends another underscore to it, forming a name `__foo` at runtime. This use has occurred for ages since Bruce Mitchener invented the `$` prefix concept, here's one of the earlier uses at e.g. a commit from June 2014, https://github.com/emscripten-core/emscripten/commit/946a27ee58ddd6cdcfcc896fea0f8187e2263795#diff-c7158ec647818499b7c892fd35ce45970083449a24806c3b25355b7c4046760fR1197 . 

To my understanding that is a good way to mark symbols internal, not sure why it had to change. The rationale for that use is that in C standard, [all functions that begin with an underscore at file scope are reserved names](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n1256.pdf):

```
All identifiers that begin with an underscore are always reserved
for use as identifiers with file scope in both the ordinary and tag name spaces
```

Hence prepending JS functions with a `_` should guarantee no conflict with either user C/C++ symbols or user JS symbols, but it will conveniently also serve as documentation to make external JS code developer pause "this symbol starts with a double underscore __ so looks very internal, don't call it from JS liberally".

In #11738 and before, this convention was turned on its head: what used to be internal symbols are now public symbols, and vice versa. I don't understand why that needed to be changed - to my view, the convention was:
 - `$foo` (becomes `foo`)=public to user JS, not accessible from C,
 - `_foo` (becomes `__foo`)=internal impl to JS system libs, don't call from user JS or C,
 - `foo` (becomes `_foo`)=public to user C and JS

and that was workable in my experience. The only inconvenience with that method was that the unprefixed function names that were public to both JS and C at the same time would have a single underscore in them, making it a little bit more inconvenient to type to access them from JS. But this would be semi-ok, since often times the ABI would want to be different anyways (JS strings/arrays vs Wasm C array/string marshalling), so if one really wanted to call the function conveniently from JS, one would write a `$foo` variant that would do the marshalling before calling `_foo`.

Now after #11738 et al. we seem to have
 - `$foo`=private from user JS (but still pollutes user JS namespace), not accessible from C,
 - `_foo`=should not be used?
 - `foo`=public to user C and JS

which seems it gives more problems with no benefit. It results in inability to distinguish between actually public and private JS symbols (`$FS` vs `$wasmTableMirror`, why is accessing `FS` ok but `wasmTableMirror` not?). Private JS symbols that users should not call are polluting user JS namespace (and still has the same inconvenience than the above method with symbols public to both C and JS, no improvement there). It does avoid producing new `_foo` symbols to C scope, but those would be reserved names by the spec anyways.

I wish we could go back to the old convention which had more benefits, but I don't know if back-and-forth will be more confusing. Although this is not a big issue for me, since I know "by heart" which JS symbols are ok to call and which are not, so I can also change this to `$wasmTableMirror` if that's preferable. lmk.

Reply:
I think I understand now.

I agree the old solution was nicer, except for the fact the it exposed JS symbols (the ones starting with `_` to C).   I think we can find a way to go back to the benefits of that system.   And we should document it somewhere when we do and I'm happy to revert/change #11738 at that time.

The one thing I would like to try to do is avoid having private JS symbols exposed to (accessible from) C.   I wonder if we can find some way to achieve that.

The problem with using `_` to mark internal JS functions is that we have plenty of C APIs that we need to implement that also have to start with `_` so we have to make them C accessible.   i.e. it would be nice to find some way to distinguish between `_jsHelper` and `_emscripten_internal_c_function_impl`.

Reply:
> The one thing I would like to try to do is avoid having private JS symbols exposed to (accessible from) C.

I don't think there is much benefit or need for that. Why is that a thing? The rationale why that is not a big deal is that if you are a JS developer, and are looking out for what you can call, you'll typically first look at a) JS documentation, b) the src/library_*.js files, and c) the generated/emitted JS code.

Then if you are a C developer, and are looking out for what one can call, you'll do primarily what C developers do: a) look at the API documentation and b) look at the system .h files in the sysroot.

That is, C developers would not go looking into JS code for what things they can call.

Under that scenario, it will really never happen that a C developer would accidentally call an internal JS function that got imported to the C namespace. They would first have to accidentally declare it in their code, and that just won't happen unless one is explicitly out to abuse things.

So ultimately not sure what is the scenario where we'd care? Those functions might get imported to the wasm namespace, but metadce will eliminate those? And namespace conflicts should not happen if people have written standards compliant C code?

In any case, I think we could entertain adding a rule that any JS side symbol that starts with a underscore should not get imported ever to wasm, or something like that? And `$` would denote a public JS API?

Reply:
Renamed to `$wasmTableMirror` for now anyways.